independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java

```diff
@@ -241,13 +239,15 @@ public Enumeration<URL> getResources(String unsanitisedName, boolean parentAlrea
         }
         //TODO: in theory resources could have been added in dev mode
         //but I don't thing this really matters for this code path
+        Set<URL> resources = new LinkedHashSet<>();
```

Comment:
It would seem that the allocation of this Set could be deferred even further, at cost of making code slightly less readable.
Up to you.. I wouldn't do it but since you're zooming into such details :)

Remember this one isn't the classloader used in production.

Reply:
This may be a little less readable.
What can be done however is to lower the size of the set at the default (16 with a factor à 75%) may be a little too much as I would assume that most of the time there'll be 1 resource (or a very few).
Also I wonder if the order is that important and using an ArrayList would be possible.

> Remember this one isn't the classloader used in production.

Yes, I can have a look at the RunnerClassLoader, maybe the same mico-optim could be done but anyway, dev classloader perf is also important.

Reply:
Unless we do some measurements, let's not guess the optimal size. For most classes - yes, 1 would be good enough but there are plenty of other resources such as `META-INF/services/xxx`, `META-INF/beans.xml`, etc.

Reply:
Reading the default ClassLoader code, they use an array of length 2 so here, using a LinkedHashSet of size 3 will do the same.

Reply:
@aloubyansky we can still lower to 8 or 6 for ex, I would expect most resources to be under 4 even for service loader and special config classes. Allocation would be a lot less as a typical application load tens of thousands of  classes.

The same can be done for `public List<ClassPathElement> getElementsWithResource(String name, boolean localOnly)` that allocate a default sized ArrayList.

> Unless we do some measurements, let's not guess the optimal size.

If we cannot guess (but usually the defaults of the JDK are OK, and they use 2), so yes, we may need to measure it.

Reply:
Sounds reasonable.

Reply:
@aloubyansky what sound reasonable ;)

Lowering to 8 (or 6) or measuring it ?

Reply:
You're right, I read too quickly it uses two enumerations that compose the resulting results.
I'll think about it and made some measurements.

Reply:
I was curious so I made a quick and dirty measurements on a very small apps using hacks inside the QuarkusClassLoader and it gives an average of 1.3 for getResources() and 1 for getElementsWithResources()

```
getResources(): IntSummaryStatistics{count=505, sum=694, min=0, average=1,374257, max=43}
getElementsWithResources():IntSummaryStatistics{count=181, sum=271, min=1, average=1,497238, max=17}
```

Reply:
TBH I'd leave this as it is. There's very low return on our time discussing this now and especially for future maintainers to try to understand this.

@loicmathieu if you'd like to get more involved in performance I'd be happy to share some puzzles with you which have the potential for high impact, we have plenty of cool things to do :)

Reply:
OK, cool, I'll let this as is.

@Sanne always interested with performance challenges ;)